extensions/data-workspace/src/test/workspaceService.test.ts

```diff
@@ -0,0 +1,211 @@
+		should.strictEqual(extension2.activationStub.notCalled, true, 'extension2.activate() should not have been called');
+		should.strictEqual(extension3.activationStub.notCalled, true, 'extension3.activate() should not have been called');
+		should.strictEqual(extension4.activationStub.calledOnce, true, 'extension4.activate() should have been called');
+		should.strictEqual(extension5.activationStub.called, true, 'extension5.activate() should have been called');
+		should.strictEqual(extension6.activationStub.notCalled, true, 'extension6.activate() should not have been called');
+		should.strictEqual(extension7.activationStub.notCalled, true, 'extension7.activate() should not have been called');
+		should.strictEqual(consoleErrorStub.calledOnce, true, 'Logger.error should be called once');
+	});
+
+	test('test getProjectProvider', async () => {
+		const extension1 = createMockExtension('ext1', true, ['csproj']);
+		const extension2 = createMockExtension('ext2', false, ['sqlproj']);
+		const extension3 = createMockExtension('ext3', false, ['dbproj']);
+		stubAllExtensions([extension1, extension2, extension3].map(ext => ext.extension));
+		const getProviderByProjectTypeStub = sinon.stub(ProjectProviderRegistry, 'getProviderByProjectType');
+		getProviderByProjectTypeStub.onFirstCall().returns(undefined);
+		getProviderByProjectTypeStub.onSecondCall().returns(createProjectProvider([
+			{
+				projectFileExtension: 'sqlproj',
+				icon: '',
+				displayName: 'test project'
+			}
+		]));
+		let provider = await service.getProjectProvider(vscode.Uri.file('abc.sqlproj'));
+		should.notStrictEqual(provider, undefined, 'Provider should be returned for sqlproj');
+		should.strictEqual(provider!.supportedProjectTypes[0].projectFileExtension, 'sqlproj');
+		should.strictEqual(extension1.activationStub.notCalled, true, 'the ext1.activate() should not have been called for sqlproj');
+		should.strictEqual(extension2.activationStub.calledOnce, true, 'the ext2.activate() should have been called once after requesting sqlproj provider');
+		should.strictEqual(extension3.activationStub.notCalled, true, 'the ext3.activate() should not have been called for sqlproj');
+
+		getProviderByProjectTypeStub.reset();
+		getProviderByProjectTypeStub.returns(createProjectProvider([{
+			projectFileExtension: 'csproj',
+			icon: '',
+			displayName: 'test cs project'
+		}]));
+		provider = await service.getProjectProvider(vscode.Uri.file('abc.csproj'));
+		should.notStrictEqual(provider, undefined, 'Provider should be returned for csproj');
+		should.strictEqual(provider!.supportedProjectTypes[0].projectFileExtension, 'csproj');
+		should.strictEqual(extension1.activationStub.notCalled, true, 'the ext1.activate() should not have been called for csproj');
+		should.strictEqual(extension2.activationStub.calledOnce, true, 'the ext2.activate() should still have been called once');
```

Comment:
shouldn't extension1 be activated instead of extension2 because this is a csproj?

Reply:
the activate method should not have been called here because the extension is already loaded.